### PR TITLE
[exporter/datadog] Support ddtags when batching logs

### PR DESCRIPTION
**Pull request by liustanley**
**Description:** <Describe what has changed.>

Fixes bug where ddtags are not set correctly when logs are batched.

**Link to tracking Issue:** #17398 

**Testing:** <Describe what testing was performed and which tests were added.>
Added unit tests for batched logs and tested locally

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

**Comment by jmichalek132** [comment]
Hi, I have a question regarding the implementation of this fix. From what I saw during my testing if don't submit any tags as part of [opts](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/17399/files#diff-0712484244af9215bc1fbbf1ef6f2543b91527d81742c102e8c3dbe7d59357dbR79) than the backend actually looks at tags that are in the payload.
Would it be better than just to iterate over log items in the payload if we need to inject extra tags from the configuration of the exporter instead of calling submitLog in a loop?

File: exporter/datadogexporter/internal/logs/sender.go
Comment (dineshg13) [reply exporter/datadogexporter/internal/logs/sender.go:91]: Can you please add some comments, why we are doing this way ?

**Review by dineshg13 — CHANGES_REQUESTED** [review]

**Comment by liustanley** [comment]
> Hi, I have a question regarding the implementation of this fix. From what I saw during my testing if don't submit any tags as part of [opts](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/17399/files#diff-0712484244af9215bc1fbbf1ef6f2543b91527d81742c102e8c3dbe7d59357dbR79) than the backend actually looks at tags that are in the payload. Would it be better than just to iterate over log items in the payload if we need to inject extra tags from the configuration of the exporter instead of calling submitLog in a loop?

@jmichalek132 Thanks for sharing this insight. I tried implementing the fix in this way and it works for tags defined in the log message, but it doesn't always include tags such as `otel_source:datadog_exporter` and `os.type:darwin` that are added in the Datadog Exporter. This happens due to a behavior in the Datadog backend, so I think setting `opts` would work better with our backend.

File: exporter/datadogexporter/internal/logs/sender.go
Comment (liustanley) [reply exporter/datadogexporter/internal/logs/sender.go:107]: I think this block needs to be separate from the `prevDdtags == tags` condition. The last loop iteration always needs to run `SubmitLog` to submit the last log(s).

File: exporter/datadogexporter/internal/logs/sender.go
Comment (gbbr) [inline exporter/datadogexporter/internal/logs/sender.go:69]: Can you create a group? And maybe make the variable names a bit more friendly as per our [coding guidelines](https://datadoghq.atlassian.net/wiki/spaces/OTEL/pages/2580119947/Go+Coding+Guidelines#Naming-things) ?

e.g. 
```suggestion
	var (
		prevtags string // ... 
		batch []datadogV2.HTTPLogItem // ... 
	)
```
Can you please document them too?

File: exporter/datadogexporter/internal/logs/sender.go
Comment (gbbr) [inline exporter/datadogexporter/internal/logs/sender.go:76]: I would either rename `payload` to `logItems` (which makes complete sense). Or rename `logItem` to `p`.

File: exporter/datadogexporter/internal/logs/sender.go
Comment (gbbr) [inline exporter/datadogexporter/internal/logs/sender.go:85]: Even though `fmt.Sprint` seems like a good convenience method, it's bloody expensive and is a top offender on profiles. For processing potentially high throughput log lines, I recommend avoiding it, especially for a basic concatenation such as this one.

File: exporter/datadogexporter/internal/logs/sender.go
Comment (gbbr) [inline exporter/datadogexporter/internal/logs/sender.go:78]: This does not add any value. What the comment says is explicitly clear by reading the line below it. Please remove it.

File: exporter/datadogexporter/internal/logs/sender.go
Comment (gbbr) [inline exporter/datadogexporter/internal/logs/sender.go:84]: When will a freshly initialised `opts` (5 lines above) ever contain a non-nil Ddtags?

File: exporter/datadogexporter/internal/logs/sender.go
Comment (gbbr) [inline exporter/datadogexporter/internal/logs/sender.go:90]: Please put this line inside the `if` statement, as per our [coding guidelines](https://datadoghq.atlassian.net/wiki/spaces/OTEL/pages/2580119947/Go+Coding+Guidelines).

File: exporter/datadogexporter/internal/logs/sender.go
Comment (gbbr) [inline exporter/datadogexporter/internal/logs/sender.go:98]: Please create a separate helper function for this logic (which is repeated below too). And don't create one `opts` at the top of the function and keep overriding its DdTags field. Instead, create it in that helper function. The helper function should receive the "tags" (to be used for opts) and the "bath" to be written.

**Review by gbbr — CHANGES_REQUESTED** [review]
The code is very difficult to follow and needs to be simplified. I'm happy to help work through this. Left an initial round of comments.

File: exporter/datadogexporter/internal/logs/sender.go
Comment (gbbr) [inline exporter/datadogexporter/internal/logs/sender.go:103]: Can we find a way to do it after the loop exits without the `if i == len(payload-1)` which basically means the same thing.

**Comment by gbbr** [comment]
Have you considered having a map from tags to log items? You could then place each log item at the corresponding key and after the loop finishes flush all batches. 

This would also cover the case where the same ddtags items are not consecutive.

File: exporter/datadogexporter/internal/logs/sender.go
Comment (liustanley) [reply exporter/datadogexporter/internal/logs/sender.go:84]: Add unit test for this before deleting

**Comment by liustanley** [comment]
> Have you considered having a map from tags to log items? You could then place each log item at the corresponding key and after the loop finishes flush all batches.
> 
> This would also cover the case where the same ddtags items are not consecutive.

@gbbr I tried using a map in this way, but it results in logs being sent out of order as batch order takes priority instead. I think the current approach is the most consistent way to optimize log batching.

**Review by dineshg13 — APPROVED** [review]
LTGM.

File: exporter/datadogexporter/internal/logs/sender.go
Comment (gbbr) [reply exporter/datadogexporter/internal/logs/sender.go:72]: You did not address this one.

**Review by gbbr — APPROVED** [review]
Thanks Stanley! LGTM as far as code readability goes. I'll let @dineshg13 review the logic. I hope you agree this looks better.

P.S. You resolved a comment that you did not address (minor)

**Review by mx-psi — APPROVED** [review]

**Pull request merged**
